core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1066 @@
+    opnd_t opnd = opnd_create_base_disp(reg, DR_REG_NULL, 0, imm, OPSZ_8);
+    instr_set_dst(out, idx, opnd);
+    return true;
+}
+/**
+ * Array of operand decode functions indexed by riscv64_fld_t.
+ *
+ * NOTE: After benchmarking, perhaps this could be placed in the same section as
+ *       instr_infos and trie?
+ */
+opnd_dec_func_t opnd_decoders[] = {
+    [RISCV64_FLD_NONE] = decode_none_opnd,
+    [RISCV64_FLD_RD] = decode_rd_opnd,
+    [RISCV64_FLD_RDFP] = decode_rdfp_opnd,
+    [RISCV64_FLD_RS1] = decode_rs1_opnd,
+    [RISCV64_FLD_RS1FP] = decode_rs1fp_opnd,
+    [RISCV64_FLD_BASE] = decode_base_opnd,
+    [RISCV64_FLD_RS2] = decode_rs2_opnd,
+    [RISCV64_FLD_RS2FP] = decode_rs2fp_opnd,
+    [RISCV64_FLD_RS3] = decode_rs3_opnd,
+    [RISCV64_FLD_FM] = decode_fm_opnd,
+    [RISCV64_FLD_PRED] = decode_pred_opnd,
+    [RISCV64_FLD_SUCC] = decode_succ_opnd,
+    [RISCV64_FLD_AQRL] = decode_aqrl_opnd,
+    [RISCV64_FLD_CSR] = decode_csr_opnd,
+    [RISCV64_FLD_RM] = decode_rm_opnd,
+    [RISCV64_FLD_SHAMT] = decode_shamt_opnd,
+    [RISCV64_FLD_SHAMT5] = decode_shamt5_opnd,
+    [RISCV64_FLD_SHAMT6] = decode_shamt6_opnd,
+    [RISCV64_FLD_I_IMM] = decode_i_imm_opnd,
+    [RISCV64_FLD_S_IMM] = decode_s_imm_opnd,
+    [RISCV64_FLD_B_IMM] = decode_b_imm_opnd,
+    [RISCV64_FLD_U_IMM] = decode_u_imm_opnd,
+    [RISCV64_FLD_J_IMM] = decode_j_imm_opnd,
+    [RISCV64_FLD_CRD] = decode_crd_opnd,
+    [RISCV64_FLD_CRDFP] = decode_crdfp_opnd,
+    [RISCV64_FLD_CRS1] = decode_crs1_opnd,
+    [RISCV64_FLD_CRS2] = decode_crs2_opnd,
+    [RISCV64_FLD_CRS2FP] = decode_crs2fp_opnd,
+    [RISCV64_FLD_CRD_] = decode_crd__opnd,
+    [RISCV64_FLD_CRD_FP] = decode_crd_fp_opnd,
+    [RISCV64_FLD_CRS1_] = decode_crs1__opnd,
+    [RISCV64_FLD_CRS2_] = decode_crs2__opnd,
+    [RISCV64_FLD_CRS2_FP] = decode_crs2_fp_opnd,
+    [RISCV64_FLD_CRD__] = decode_crd___opnd,
+    [RISCV64_FLD_CSHAMT] = decode_cshamt_opnd,
+    [RISCV64_FLD_CSR_IMM] = decode_csr_imm_opnd,
+    [RISCV64_FLD_CADDI16SP_IMM] = decode_caddi16sp_imm_opnd,
+    [RISCV64_FLD_CLWSP_IMM] = decode_clwsp_imm_opnd,
+    [RISCV64_FLD_CLDSP_IMM] = decode_cldsp_imm_opnd,
+    [RISCV64_FLD_CLUI_IMM] = decode_clui_imm_opnd,
+    [RISCV64_FLD_CSWSP_IMM] = decode_cswsp_imm_opnd,
+    [RISCV64_FLD_CSDSP_IMM] = decode_csdsp_imm_opnd,
+    [RISCV64_FLD_CIW_IMM] = decode_ciw_imm_opnd,
+    [RISCV64_FLD_CLW_IMM] = decode_clw_imm_opnd,
+    [RISCV64_FLD_CLD_IMM] = decode_cld_imm_opnd,
+    [RISCV64_FLD_CSW_IMM] = decode_csw_imm_opnd,
+    [RISCV64_FLD_CSD_IMM] = decode_csd_imm_opnd,
+    [RISCV64_FLD_CIMM5] = decode_cimm5_opnd,
+    [RISCV64_FLD_CB_IMM] = decode_cb_imm_opnd,
+    [RISCV64_FLD_CJ_IMM] = decode_cj_imm_opnd,
+    [RISCV64_FLD_V_L_RS1_DISP] = decode_v_l_rs1_disp_opnd,
+    [RISCV64_FLD_V_S_RS1_DISP] = decode_v_s_rs1_disp_opnd,
+};
+
+static inline rv_instr_info_t *
+match_op_0(int funct, bool rv32, bool rv64)
+{
+    switch (funct) {
+    case 0: return &instr_infos[OP_c_addi4spn];
```

Comment:
I'm not sure where these case values are coming from?

Reply:
They were crafted using tables 16.5–16.7 in the spec. The compressed instruction set is rather small and there are many which alias other instructions so it was simpler to decode compressed instructions explicitly than include them in the search trie created for uncompressed instructions.

Reply:
Done. I've renamed those functions to better match the spec tables.